libsignal-service-actix/src/push_service.rs

```diff
@@ -182,6 +183,70 @@ impl PushService for AwcPushService {
         ))
     }
 
+    async fn post_to_cdn0<'s, C: std::io::Read + Send + 's>(
+        &mut self,
+        path: &str,
+        value: &[(&str, &str)],
+        file: Option<(&str, &'s mut C)>,
+    ) -> Result<(), ServiceError> {
+        let url = Url::parse(&self.cfg.cdn_urls[&0])
+            .expect("valid cdn base url")
+            .join(path)
+            .expect("valid CDN path");
+
+        log::debug!("AwcPushService::post_to_cdn({:?})", url);
+        let mut client = self.client.post(url.as_str());
+
+        let mut form = multipart::Form::default();
+
+        for (k, v) in value {
+            form.add_text(*k, *v);
+        }
+
+        if let Some((filename, file)) = file {
+            // XXX Actix doesn't cope with none-'static lifetimes
+            // https://docs.rs/actix-web/3.2.0/actix_web/body/enum.Body.html
+            let mut buf = Vec::new();
+            file.read_to_end(&mut buf)
+                .expect("infallible Read instance");
+            form.add_reader(filename, std::io::Cursor::new(buf));
+            client = client.header("Content-Type", "application/octet-stream");
+        }
+
+        let content_type = form.content_type();
+
+        // XXX Amazon S3 needs the Content-Length, but we don't know it without depleting the whole
+        // stream.  Sadly, Content-Length != contents.len(), but should include the whole form.
+        let mut body = multipart::Body::from(form);
+        let mut body_contents = vec![];
+        use actix_http::body::MessageBody;
+        use futures::stream::StreamExt;
+        let mut stream = futures::stream::poll_fn(move |ctx| {
+            let body = std::pin::Pin::new(&mut body);
```

Comment:
maybe `pin_mut!(body)` instead?

Reply:
Something very weird happens if I try:

```
error[E0507]: cannot move out of `body`, a captured variable in an `FnMut` closure
   --> libsignal-service-actix/src/push_service.rs:225:13
    |
220 |         let mut body = multipart::Body::from(form);
    |             -------- captured outer variable
...
225 |             futures::pin_mut!(body);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^
    |             |
    |             move occurs because `body` has type `actix_multipart_rfc7578::client::multipart::Body<'_>`, which does not implement the `Copy` trait
    |             help: consider borrowing here: `&$x`
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: variable does not need to be mutable
   --> libsignal-service-actix/src/push_service.rs:220:13
    |
220 |         let mut body = multipart::Body::from(form);
    |             ----^^^^
    |             |
    |             help: remove this `mut`
    |
    = note: `#[warn(unused_mut)]` on by default
```

Reply:
You can ignore my comment then, thanks for trying.